### PR TITLE
Support node and edge deletion from graph

### PR DESCRIPTION
**Pull request by swilly22**

### CODE REVIEW

File: src/graph/graph.c [hunk 1]
```diff
@@ -60,6 +66,86 @@ void _Graph_ResizeNodes(Graph *g, size_t n) {
...
+void _Graph_RemoveNodeFromBlockchain(Graph *g, int id, int *replacementID) {
+    // Get the block in which removed node resides.
+    NodeBlock *removedNodeBlock = GRAPH_GET_NODE_BLOCK(g, id);
+    // Get node position within its block.
+    int nodeBlockIdx = GRAPH_NODE_POSITION_WITHIN_BLOCK(id);
+    // Get the newest node in the graph.
+    GrB_Index newestNodeID = g->node_count-1;
+    Node *newestNode = Graph_GetNode(g, newestNodeID);
+    // Replace removed node with newest node.
+    removedNodeBlock->nodes[nodeBlockIdx] = *newestNode;
```
Comment (jeffreylovitz) [inline src/graph/graph.c]: Won't this cause later attempts to access the newest node in the `NodeBlock` to fail? Its ID will still point to the latest block and offset.

Comment (swilly22) [reply src/graph/graph.c]: No, as IDs are internal, and are computed by the position of an entity within a matrix,
and so a node which was at row I and got moved to row J will have its ID update at some point to J,
we're updating node IDs lazily.

File: src/execution_plan/ops/op_delete.c [hunk 2]
```diff
@@ -1,253 +1,146 @@
...
+
+/* Forward declarations. */
+void _LocateEntities(OpDelete *op_delete, QueryGraph *graph, AST_DeleteNode *ast_delete_node);
+
+OpBase* NewDeleteOp(AST_DeleteNode *ast_delete_node, QueryGraph *qg, Graph *g, ResultSet *result_set) {
+    return (OpBase*)_NewDeleteOp(ast_delete_node, qg, g, result_set);
+}
+
+OpDelete* _NewDeleteOp(AST_DeleteNode *ast_delete_node, QueryGraph *qg, Graph *g, ResultSet *result_set) {
+    OpDelete *op_delete = (OpDelete*) calloc(1, sizeof(OpDelete));
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_delete.c]: Every member of the struct is explicitly set, so malloc should be fine.

File: src/graph/graph.c [hunk 3]
```diff
@@ -57,9 +63,72 @@ void _Graph_ResizeNodes(Graph *g, size_t n) {
...
+    NodeBlock *destNodeBlock = GRAPH_GET_NODE_BLOCK(g, dest);
+    
+    // Get node position within its block.
+    int destNodeBlockIdx = GRAPH_NODE_POSITION_WITHIN_BLOCK(dest);
+
+    // Get the src node in the graph.
+    Node *srcNode = Graph_GetNode(g, src);
+
+    // Replace dest node with src node.
+    destNodeBlock->nodes[destNodeBlockIdx] = *srcNode;
```
Comment (jeffreylovitz) [inline src/graph/graph.c]: I think that the `Node`'s `incoming_edges` and `outgoing_edges` Vectors only get populated for QueryGraph members, but if I am incorrect then they will be broken here.

(Should we consider simplifying that struct?)

File: src/graph/graph.c [hunk 4]
```diff
@@ -109,32 +178,30 @@ Graph *Graph_Get(RedisModuleCtx *ctx, RedisModuleString *graph_name) {
...
-        // Reset iterator for caller.
-        NodeIterator_Reset(node_it);
     }
 
-    g->node_count += n;
-
-    if(it != NULL) *it = node_it;
-    else NodeIterator_Free(node_it);
+    if(it != NULL) {
+        *it = NodeIterator_New(GRAPH_ACTIVE_BLOCK(g),
```
Comment (jeffreylovitz) [inline src/graph/graph.c]: I think that if n is large enough that we are completely populating one or more additional blocks, this iterator will start in a later block than it ought to.

We can avoid this by making the argument on line 202 `g->node_count + n` and updating `g->node_count` after this call.

File: src/graph/graph.c [hunk 5]
```diff
@@ -170,6 +237,146 @@ Node* Graph_GetNode(const Graph *g, int id) {
...
+    typedef struct {
+        int nodeID;         // Node being deleted.
+        int replacementID;  // Node taking over.
+        bool delete;        // No need to replace, simply delete.
+    } Replacement;
+
+    Replacement *replacements = malloc(sizeof(Replacement) * IDCount);
+
+    /* Allocate replacement candidates. */
+    for(int i = 0; i < IDCount; i++) {
```
Comment (jeffreylovitz) [inline src/graph/graph.c]: I feel like this approach is more elaborate than it needs to be. Can we just:

```
Sort the IDs array in descending order
int innermost = IDCount - 1;
for (i = 0; i < innermost; i ++) {
  for (j = IDs[i] - 1; j > IDs[i + 1]; j --) {
    replace node at IDs[innermost] with node at j, update matrices
    innermost --;
    if (innermost <= i) {
      max_remaining = j;
      break;
    }
  }
}

resize matrix to max_remaining
```
I am likely overlooking some edge cases and getting some offsets wrong, but generally I feel like if we sort the nodes to be deleted, we can quickly tell which higher-ID nodes are safe candidates to use as replacements.
Then we can swap them immediately, and continue doing so until there are no to-be-deleted IDs less than the updated node count.
At that point, we can resize the matrix down to the ID of the last replacement node we swapped in.

File: src/graph/graph.c (hunk 3 above)
Comment (swilly22) [reply src/graph/graph.c]: indeed edges are only used within the context of a query graph, we should create a query node and a query edge, and update the node object.

File: src/graph/graph.c (hunk 4 above)
Comment (swilly22) [reply src/graph/graph.c]: You're absolutely right! nice catch

File: src/execution_plan/ops/op_delete.c (hunk 2 above)
Comment (swilly22) [reply src/execution_plan/ops/op_delete.c]: 👍

File: src/graph/graph.c (hunk 5 above)
Comment (swilly22) [reply src/graph/graph.c]: At first I thought about sorting which will defiantly make this code clearer, once we have a decent sorting algorithm in place there are additional areas in the code which can benefit from it.
I'll keep this as is for now and return to it once we find an optimised sorting algorithm.

**Pull request merged**